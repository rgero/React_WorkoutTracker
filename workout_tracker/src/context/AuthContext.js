import AsyncStorage from '@react-native-async-storage/async-storage';

import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const authReducer = (state, action) => {
    switch (action.type)
    {
        case "addError":
            return { ...state, errorMessage: action.payload};
        case "storeToken":
            return { errorMessage: "", token: action.payload};
        case "clearErrorMessage":
            return {...state, errorMessage: ""}
        case "signout":
            return { token: null, errorMessage: ""}
        default:
            return state;
    }
};

const tryLocalSignin = dispatch => async () => {
    const token = await AsyncStorage.getItem('token');
    if (token)
    {
        dispatch({ type: 'storeToken', payload: token})
    }
}

const clearErrorMessage = dispatch => () => {
    dispatch( {type: 'clearErrorMessage' })
}

const signUp = dispatch => {
    return async ({ email, displayName, password }) => {
        try {
            const response = await trackerAPI.post('/signup', { email, displayName, password });
            
            // Storing the Token
            await AsyncStorage.setItem('token', response.data.token);
            dispatch({type: "storeToken", payload: response.data.token})
        } catch (err) {
            console.log(err.message);
            dispatch({ type: 'addError', payload: "Sign-up failed. Please try again"})
        }
    };
}

const signIn = (dispatch) => {
    return async ({ email, password }) => {
        try {
            const response = await trackerAPI.post('/signin', { email, password });
            
            // Storing the Token
            await AsyncStorage.setItem('token', response.data.token);
            dispatch({type: "storeToken", payload: response.data.token})
        } catch (err) {
            console.log(err.message);
            dispatch({ type: 'addError', payload: "Login failed. Please try again"})
        }
    };
}

const signOut = dispatch => async () => {
    try {
        await AsyncStorage.removeItem('token');
        dispatch({type: "signout"})
    } catch (err)
    {
        console.log(err);
    }
}

export const {Provider, Context } = createDataContext(authReducer, 
                                                      {clearErrorMessage, signUp, signIn, signOut, tryLocalSignin}, 
                                                      {
                                                        token: null,
                                                        errorMessage: ""
                                                      }
                                                    )