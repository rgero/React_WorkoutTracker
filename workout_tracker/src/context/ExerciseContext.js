import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const exerciseReducer = (state, action) => {
    switch(action.type)
    {
        case 'fetch':
            return action.payload;
        default:
            return state;
    }
};

const fetchExercises = dispatch => async () => {
    const response = await trackerAPI.get('/exercises');
    dispatch({type: 'fetch', payload: response.data})
};

const fetchExerciseByID = dispatch => async (targetID) => {
    const response = await trackerAPI.get(`/exercises/${targetID}`);
    dispatch({type: 'fetch', payload: response.data})
};


export const {Provider, Context} = createDataContext(
    exerciseReducer,
    {fetchExercises, fetchExerciseByID},
    []
);