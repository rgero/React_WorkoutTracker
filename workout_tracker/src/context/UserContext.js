import AsyncStorage from '@react-native-async-storage/async-storage';

import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const userReducer = (state, action) => {
    switch(action.type)
    {
        case 'fetch':
            return action.payload;
        default:
            return state;
    }
};

const getUser = dispatch => async () => {
    const token = await AsyncStorage.getItem('token');
    if (token)
    {
        const response = await trackerAPI.get('/users');
        dispatch({type: 'fetch', payload: response.data})
    }
}

export const {Provider, Context} = createDataContext(
    userReducer,
    {getUser},
    []
);