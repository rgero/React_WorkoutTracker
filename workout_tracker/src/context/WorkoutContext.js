import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const workoutReducer = (state, action) => {
    switch(action.type)
    {
        case 'fetch':
            return action.payload;
        case "create":
            return state.concat(action.payload);
        case "edit":
            return state.map( (workout) => {
                if (workout._id === action._id) {
                    return {
                        ...workout,
                        ...action.update
                    }
                } else {
                    return workout;
                }
            })
        case "delete":
            return state.filter( (testWorkout) => {
                return action.id !== testWorkout._id;
            });
        default:
            return state;
    }
};

const fetchWorkouts = dispatch => async () => {
    const response = await trackerAPI.get('/workouts');
    dispatch({type: 'fetch', payload: response.data})
};

const createWorkout = dispatch => async (workout) => 
{
    const response = await trackerAPI.post('/workouts', workout)
    dispatch({type: "create", payload: response.data});
};

const updateWorkout = dispatch => async (workout) =>
{
    const response = await trackerAPI.put(`/workouts/${workout._id}`, workout);
    dispatch({type: "edit", payload: response.data});
}

const deleteWorkout = dispatch => async (targetID) => 
{
    const response = await trackerAPI.delete(`/workouts/${targetID}`)
    dispatch({type: "delete", id: targetID});
}

export const {Provider, Context} = createDataContext(
    workoutReducer,
    {fetchWorkouts, createWorkout, deleteWorkout, updateWorkout},
    []
);