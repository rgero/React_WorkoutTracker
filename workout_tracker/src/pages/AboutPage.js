import Container from 'react-bootstrap/Container';

function AboutPage() {
    return (
        <Container>
            <a href="https://www.flaticon.com/free-icons/gym" title="gym icons">Gym icons created by Freepik - Flaticon</a>
        </Container>
    );
  }
  
  export default AboutPage;