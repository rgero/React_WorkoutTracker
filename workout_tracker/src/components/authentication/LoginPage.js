import React, { useContext, useEffect, useState } from 'react';
import Container from 'react-bootstrap/Container';
import { Navigate } from "react-router-dom";
import {Context as AuthContext} from '../../context/AuthContext';

import { LoginForm } from './LoginForm';


export const LoginPage = ()=> 
{
    const {state, signIn, clearErrorMessage, tryLocalSignin} = useContext(AuthContext);

    useEffect(()=> {
        tryLocalSignin();
    }, [])

    if (state.token)
    {
        clearErrorMessage();
        return (<Navigate replace to="/dashboard" />);
    } else 
    {
        return (
            <Container fluid="md">
                <h1>Login</h1>
                <LoginForm
                    errorMessage={state.errorMessage}
                    onSubmit={({email, password})=> signIn({email, password})}
                />
            </Container>
        )
    }
}