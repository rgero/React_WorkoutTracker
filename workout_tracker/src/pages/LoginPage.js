import React, { useContext, useEffect } from 'react';
import Alert from 'react-bootstrap/Alert';
import Container from 'react-bootstrap/Container';
import { Navigate } from "react-router-dom";
import {Context as AuthContext} from '../context/AuthContext';
import { LoginForm } from '../components/authentication/LoginForm';



export const LoginPage = ()=> 
{
    const {state, signIn, clearErrorMessage, tryLocalSignin} = useContext(AuthContext);

    useEffect(()=> {
        const tryToSignIn = async () => {
            await tryLocalSignin();
            if (state.token)
            {
                clearErrorMessage();
            }
        }

        tryToSignIn();
    }, [])

    if (state.token)
    {
        return (<Navigate replace to="/dashboard" />);
    } else 
    {
        return (
            <Container fluid="md">
                <h1>Log in to the Workout Tracker</h1>
                <LoginForm
                    errorMessage={state.errorMessage}
                    onSubmit={({email, password})=> signIn({email, password})}
                />
                { state.errorMessage ? (
                    <Container className="pt-4">
                        <Alert variant='danger'>Error: {state.errorMessage}</Alert>
                    </Container>
                ) : null
            }
            </Container>
        )
    }
}