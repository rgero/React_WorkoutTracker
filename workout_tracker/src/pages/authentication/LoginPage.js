import React, { useContext, useEffect } from 'react';
import { Navigate } from "react-router-dom";

import Alert from 'react-bootstrap/Alert';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';

import {Context as AuthContext} from '../../context/AuthContext';
import { LoginForm } from '../../components/authentication/LoginForm';

export const LoginPage = ()=> 
{
    const {state, signIn, clearErrorMessage, tryLocalSignin} = useContext(AuthContext);

    useEffect(()=> {
        const tryToSignIn = async () => {
            await tryLocalSignin();
            if (state.token)
            {
                clearErrorMessage();
            }
        }

        tryToSignIn();
    }, [])

    if (state.token)
    {
        return (<Navigate replace to="/dashboard" />);
    } else 
    {
        return (
            <Container>
                <Row>
                    <Col md={8} className="d-none d-lg-block">
                        <img src="/images/dumbbells.jpg" width="100%" />
                    </Col>
                    <Col xs={10} md={4}>
                        <LoginForm
                            onSubmit={({email, password})=> signIn({email, password})}
                        />
                        { state.errorMessage ? (
                                <Container className="pt-4">
                                    <Alert variant='danger'>Error: {state.errorMessage}</Alert>
                                </Container>
                            ) : null
                        }
                    </Col>
                </Row>
            </Container>
        )
    }
}

export default LoginPage;