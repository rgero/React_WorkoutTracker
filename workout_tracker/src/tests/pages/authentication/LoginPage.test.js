import React from 'react';
import TestRenderer from 'react-test-renderer';
import reactRouterDom from 'react-router-dom';
import { act } from 'react-dom/test-utils';

import {Context as AuthContext} from '../../../context/AuthContext';
import { LoginPage } from '../../../../src/pages/LoginPage';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
}));

test('Rendering the Login Page', async ()=> {

    const testContextValues = {
        state: {},
        signIn: jest.fn(),
        clearErrorMessage: jest.fn(),
        tryLocalSignin: jest.fn()
    }

    let element;
    await act(async () => {
        element = new TestRenderer.create(
            <AuthContext.Provider value={testContextValues}>
                <LoginPage/>
            </AuthContext.Provider>
        );
    })
    
    expect(element.toJSON()).toMatchSnapshot();
})

test("Login Token Presented, redirect detected", async ()=> {
    const testContextValues = {
        state: { token: "asinaidnsdin"},
        signIn: jest.fn(),
        clearErrorMessage: jest.fn(),
        tryLocalSignin: jest.fn()
    }

    let mockNavigate = jest.spyOn(reactRouterDom, "Navigate").mockImplementation();

    let element;
    await act(async () => {
        element = new TestRenderer.create(
            <AuthContext.Provider value={testContextValues}>
                <LoginPage/>
            </AuthContext.Provider>
        );
    })
    
    expect(mockNavigate.mock.calls).toEqual([[{"replace": true, "to": "/dashboard"}, {}]]);
})