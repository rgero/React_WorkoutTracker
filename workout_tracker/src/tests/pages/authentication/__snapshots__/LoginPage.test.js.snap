// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering the Login Page 1`] = `
<div
  className="pt-2 container-md"
>
  <div
    className="row"
  >
    <div
      className="col-8"
    >
      <img
        src="/images/dumbbells.jpg"
        width={800}
      />
    </div>
    <div
      className="col-lg-4 col"
    >
      <div
        className="container-md"
      >
        <div
          className="justify-content-md-center row"
        >
          <div
            className="col"
          >
            <h3>
              Log in
            </h3>
            <form
              className=""
              onSubmit={[Function]}
            >
              <div
                className="mb-3"
              >
                <label
                  className="form-label"
                  htmlFor="formEmail"
                >
                  E-mail
                </label>
                <input
                  aria-label="email"
                  className="form-control"
                  id="formEmail"
                  onChange={[Function]}
                  type="email"
                  value=""
                />
              </div>
              <div
                className="mb-3"
              >
                <label
                  className="form-label"
                  htmlFor="formPassword"
                >
                  Password
                </label>
                <input
                  aria-label="password"
                  className="form-control"
                  id="formPassword"
                  onChange={[Function]}
                  type="password"
                  value=""
                />
              </div>
              <button
                className="btn btn-outline-secondary"
                disabled={false}
                type="submit"
              >
                Log In
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
